internal/host/bruteforce/bruteforce.go

```diff
@@ -0,0 +1,200 @@
+	AnalyzeResponse(attempt bruteforce.AttemptInfo) *bruteforce.ResultInfo
+	StandardPorts() []int
+}
+
+type BruteforceEngine struct {
+	Library BruteforceLibrary
+}
+
+func (be *BruteforceEngine) Run(ctx context.Context, subStep map[string]interface{}, target string, config *bruteforce.BruteForceRunConfig) ([]*bruteforce.AttemptInfo, []string) {
+	var attempts []*bruteforce.AttemptInfo
+	errors := []string{}
+
+	// If target is not a socket, use Module standard ports
+	var ports []int
+	host, port, err := net.SplitHostPort(target)
+	if err != nil {
+		host = target
+		ports = be.Library.StandardPorts()
+	} else {
+		intPort, err := strconv.Atoi(port)
+		if err != nil {
+			errors = append(errors, err.Error())
+			return nil, errors
+		}
+		ports = []int{intPort}
+	}
+
+	for _, port := range ports {
+		for i := 0; i < config.Retries; i++ {
+			attempt, errs := be.Library.BruteForce(host, port, config.Timeout, subStep)
+			if err != nil {
+				errors = append(errors, errs...)
+			}
+			// Check SuccessfulOnly config
+			if attempt.Result.Login || !config.SuccessfulOnly {
+				attempts = append(attempts, attempt)
+			}
+			// Break if credentials successful
+			if attempt.Result.Login {
+				break
+			}
+			// Sleep if ratelimit detected
+			if attempt.Result.Ratelimit {
+				time.Sleep(time.Duration(config.Sleep) * time.Second)
+			}
+		}
+	}
+
+	return attempts, errors
+}
+
+func (be *BruteforceEngine) gatherAttemptStatistics(attempts []*bruteforce.AttemptInfo, usernames []string, passwords []string, config *bruteforce.BruteForceRunConfig) *bruteforce.StatisticsInfo {
+	stats := &bruteforce.StatisticsInfo{
+		NumUsernames: len(usernames),
+		NumPasswords: len(passwords),
+	}
+	for _, attempt := range attempts {
+		if attempt.Result.Login {
+			stats.NumSuccessful++
+		} else {
+			stats.NumFailed++
+		}
+	}
+	if config.SuccessfulOnly {
+		numUsernames := len(usernames)
+		if numUsernames == 0 {
+			numUsernames = 1
+		}
+		stats.NumFailed = (numUsernames * len(passwords) * config.Retries * len(be.Library.StandardPorts())) - stats.NumSuccessful
+	}
+	return stats
+}
+
+func BruteForceAttack(ctx context.Context, targets []string, module string, usernames []string, passwords []string, config *bruteforce.BruteForceRunConfig) (*bruteforce.BruteForceReport, error) {
+	resources := bruteforce.BruteForceReport{}
+	errors := []string{}
+
+	var engine *BruteforceEngine
+	switch module {
+	case "SSH":
+		engine = &BruteforceEngine{
+			Library: &modules.SSHLibrary{},
+		}
+	case "TELNET":
+		engine = &BruteforceEngine{
+			Library: &modules.TelnetLibrary{},
+		}
+	default:
+		return &resources, fmt.Errorf("unsupported module: %s", module)
+	}
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+
+	// Run brute force for each target/username/password combination
+	var bruteForceResults []*bruteforce.BruteForceAttempt
+	for _, target := range targets {
+		var attempts []*bruteforce.AttemptInfo
+
+		for _, password := range passwords {
+			if len(usernames) == 0 {
```

Comment:
Why are we checking the length of usernames on every iteration of passwords? We should know which option we fall in up front.

Reply:
I would recommend pre-computing your username/password pairs and storing that as a struct in a queue that you can just iterate over. It will make it easier to thread later as well.